### ISSUE TITLE
.NET Core 3.0 Angular app behind a corporate proxy may require additional environmental settings

### ISSUE THREAD

**Issue by miguellira** [aspnet/AspNetCore#14392]
### Describe the bug
Running a .NET Core 3.0 app using the default dotnet CLI Angular template behind a proxy may return unexpected results.

### To Reproduce
1. Ensure you are behind a corporate proxy
2. Create a new site via the dotnet CLI (e.g. `dotnet new angular -n HelloWorld`)
3. Run the application
4. Chromium-based browsers may respond with `ERR_UNEXPECTED_PROXY_AUTH` others will show the corporate proxy page

### Expected behavior
The application should run without error.

### Additional context
**NOTE:** I was able to workaround this by adding `localhost` to my `NO_PROXY` environment variable.

You can argue that localhost should have been in my `NO_PROXY` list to begin with, but again it worked just fine in .NET 2.x.

**Comment by javiercn** [comment]
@miguellira Thanks for contacting us.

I'm not really sure what changed here. This is likely a change in angular (we've upgraded several versions).

Does adding the following inside your profile in launchSettings.json fix the issue?
```json
"environmentVariables": {
        "NO_PROXY": "localhost",
}
```

See https://docs.microsoft.com/en-us/aspnet/core/fundamentals/environments?view=aspnetcore-3.0#development for a sample launchSettings.json profile (you should have one under /properties/ in your project root)

**Comment by miguellira** [comment]
@javiercn Yes. Modifying the `launchSettings.json` file "fixes" this.

I don't believe it is an issue with Angular or the version referenced in the .NET Core 3.0 templates. The [AngularCLIMiddleware](https://github.com/aspnet/AspNetCore/blob/master/src/Middleware/SpaServices.Extensions/src/AngularCli/AngularCliMiddleware.cs) works as expected. A random port is assigned and the ng serve command is executed. I can actually hit the Angular app directly from the browser. However, when a request to `http://localhost:5001` is made it is suppose to get proxied to the running Angular app and it is at that point that the proxy error arises.

**Comment by javiercn** [comment]
@miguellira Thanks trying it out.

@SteveSandersonMS I believe this might be caused by how the requests are being proxied from the middleware?

Could we configure the HttpClient instance I imagine the middleware is using to not use a proxy under any circumstances?

I imagine that should be safe, simple and transparent.

Alternatively we could add this to launchSettings.json or as a last resort we could simply choose to document it, as it’s not a very common case.

Thoughts?

**Comment by SteveSandersonMS** [comment]
Yes, that sounds believable. If we can make the middleware never use an external proxy (assuming it's only ever used to access a dev server running on localhost) that would be a clean fix.

**Comment by mkArtakMSFT** [comment]
We've moved this issue to the Backlog milestone. This means that it is not going to happen for the coming release. We will reassess the backlog following the current release and consider this item at that time. However, keep in mind that there are many other high priority features with which it will be competing for resources.

**Comment by mathysd** [comment]
The react template seems to have the same issue. 

It seems something has changed in the way the httpclient is working in .netcore 3.

see also my comment in:

[dotnet/corefx](https://github.com/dotnet/corefx/issues/41952#issuecomment-555460328)

**Comment by dhaselden** [comment]
I am using a corporate proxy and had the same issue with the VS 2019 Angular8 template . It looks like dotnet core 3+ has removed the system of getting your proxy settings from IE/Internet options (Interop.WinHttp.WinHttpOpen) so the "bypass proxy server for local addresses" and "Do not use proxy server for address beginning with:" no longer worked.
Setting the windows environment variables:
HTTP_PROXY
HTTPS_PROXY
NO_PROXY
worked for me. Make sure your variables are persisted and you restart Visual studio.

**Comment by dhaselden** [comment]
I am using Visual studio 2019, .net core 3.1, with Angular8 template, when using SPA modes:
spa.UseProxyToSpaDevelopmentServer
or
spa.UseAngularCliServer
I was getting a proxy message stating can't forward to localhost.

For some reason my proxy settings from "Internet Options" stopped working, I had HTTP_PROXY and HTTPS_PROXY set, but not NO_PROXY. My side by side install of Visual studio 2017 and .net core 2.1 worked fine.

Using our proxy pac file with localhost allowing direct connection didn't work. Setting "Internet Options" to bypass for local addresses or manual exceptions also didn't work.

Adding NO_PROXY environment variable with value localhost fixed my issue.

**Comment by cerberus1974** [comment]
Same issue.
This problem doesn't occur on macOS but on Windows

**Comment by cerberus1974** [comment]
Sorry. The cause was not the OS.
The problem was how to write the NO_PROXY(environment variable) .

**Comment by The-MAZZTer** [comment]
I experienced this when migrating a project from .NET Core 2.2 to 3.1.

I actually was talking to the corporate help desk for a bit since I thought it was a networking issue. Whoops.

**Comment by gregpawlowski** [comment]
I just spent a whole day attempting to get an angular project running. There are many options I tried and with angualr timing out this was very hard to pin down. Almost resorted back to running ngserve and dontet API as separate projects. Thank you for the workaround, I was surprised to learn that a proxy would affect a development environment like this. Hope this is fixed quickly or documented better.

**Comment by mumby0168** [comment]
This fix also worked for me behind a proxy, set this in my environment variables for the system so I don't have to do it for every project `NO_PROXY` -> `localhost`

**Comment by mumby0168** [comment]
Has anyone considering a PR to make this default for SPA templates?

**Comment by mumby0168** [comment]
Anyone advise whether it would be worth updating this, maybe behind a toggle as shown in this file a few times: https://github.com/dotnet/spa-templates/blob/main/src/content/Angular-CSharp/Properties/launchSettings.json

**Comment by mumby0168** [comment]
> We've significantly changed the way the SPA templates are setup for development and production to align with how other backend frameworks approach this and as a result we no longer use SpaServices as part of the development workflow.
> 
> 
> 
> While we haven't removed or obsoleted the code, we don't plan to make future changes/fixes on it and we recommend people migrate to the new approach we use in the templates.
> 
> 
> 
> We don't think this is a problem with the new approach, if that's not the case, please let us know.

Is there any links to more info on the decision and what the new middleware looks like in development?
